plotly/graph_reference.py

```diff
@@ -0,0 +1,355 @@
+    for parent_object_name in reversed(parent_object_names):
+        if parent_object_name in ARRAYS:
+            continue
+        parent_object_dict = OBJECTS[parent_object_name]
+        parent_attribute_paths = parent_object_dict['attribute_paths']
+        for path in list(attribute_paths):
+            if not _is_valid_sub_path(path, parent_attribute_paths):
+                attribute_paths.remove(path)
+
+    # We return a dict mapping paths to attributes. We also add in additional
+    # attributes if defined.
+    response = {path: utils.get_by_path(GRAPH_REFERENCE, path)
+                for path in attribute_paths}
+    response['additional_attributes'] = additional_attributes
+
+    # These are for documentation and quick lookups. They're just strings.
+    valid_names = set()
+    deprecated_names = set()
+    subplot_names = set()
+    for attributes in response.values():
+
+        for key, val in attributes.items():
+            if key not in GRAPH_REFERENCE['defs']['metaKeys']:
+                valid_names.add(key)
+                if isinstance(val, dict) and val.get('_isSubplotObj'):
+                    subplot_names.add(key)
+
+        deprecated_attributes = attributes.get('_deprecated', {})
+        for key, val in deprecated_attributes.items():
+            if key not in GRAPH_REFERENCE['defs']['metaKeys']:
+                valid_names.add(key)
+                deprecated_names.add(key)
+                if isinstance(val, dict) and val.get('_isSubplotObj'):
+                    subplot_names.add(key)
+
+    response['valid_names'] = valid_names
+    response['deprecated_names'] = deprecated_names
+    response['subplot_names'] = subplot_names
+
+    return response
+
+
+def _is_valid_sub_path(path, parent_paths):
+    """
+    Check if a sub path is valid given an iterable of parent paths.
+
+    :param (tuple[str]) path: The path that may be a sub path.
+    :param (list[tuple]) parent_paths: The known parent paths.
+    :return: (bool)
+
+    Examples:
+
+        * ('a', 'b', 'c') is a valid subpath of ('a', )
+        * ('a', 'd') is not a valid subpath of ('b', )
+        * ('a', ) is not a valid subpath of ('a', 'b')
+        * ('anything',) is a valid subpath of ()
+
+    """
+    if not parent_paths:
+        return True
+    for parent_path in parent_paths:
+        if path[:len(parent_path)] == parent_path:
+            return True
+    return False
+
+
+def _get_hr_name(object_name):
+    """Get human readable object name from reference ('hrName')."""
+    if object_name in ARRAYS:
+        return object_name  # TODO: how to handle hr name on array?
+
+    object_dict = OBJECTS[object_name]
+    meta_paths = object_dict['meta_paths']
+    for meta_path in meta_paths:
+        meta_dict = utils.get_by_path(GRAPH_REFERENCE, meta_path)
+        if meta_dict.get('_isLinkedToArray'):
+            return object_name  # TODO: how to handle hr name on array?
```

Comment:
Let me know when you resolve this, and I'll make sure that documentation logic mirrors it.